server/src/auth.js

```diff
@@ -60,97 +64,77 @@ class Auth {
   }
 
   // TODO: maybe we should write something into the user data to track open sessions/tokens
-  generate_jwt(provider, info, cb) {
-    const key = auth_key(provider, info);
-    let query = r.db(this._parent._reql_conn.metadata()._internal_db).table('users_auth').get(key);
+  generate(provider, info) {
+    const key = this.auth_key(provider, info);
+    const internal_db = r.db(this._parent._reql_conn.metadata()._internal_db);
+
+    function insert (table, row) {
+      return internal_db.table(table)
+                        .insert(row, { conflict: 'error', returnChanges: 'always' })
+                        .bracket('changes')(0)('new_val')
+    }
+
+    let query = internal_db.table('users')
+                           .get(internal_db.table('users_auth').get(key)('user_id'))
+                           .default(r.error('User not found and new user creation is disabled.'));
 
     if (this._create_new_users) {
-      query = query.default(r.uuid().do((user_id) =>
-                r.db(this._parent._reql_conn.metadata()._internal_db)
-                 .table('users_auth').insert({ id: key, user_id }, { returnChanges: true })
-                 .do((res) =>
-                   r.branch(res('inserted').eq(1),
-                     r.db(this._parent._reql_conn.metadata()._internal_db).table('users')
-                      .insert(this.new_user_row(user_id))
-                      .do((res2) =>
-                        r.branch(res2('inserted').eq(1),
-                          res2('changes')(0)('new_val'),
-                          r.error(res2('first_error'))
-                        )),
-                     r.error(res('first_error'))
-                   )
-                 )
-             ));
-    } else {
-      query = query.default(r.error('User not found and new user creation is disabled.'));
+      query = insert('users_auth', { id: key, user_id: r.uuid() })
+        .do(auth_user => insert('users', this.new_user_row(auth_user('user_id'))));
     }
 
-    this.reql_call(query).then((res) => {
-      cb(null, this._jwt_from_user(res.id, provider));
-    }).catch((err) => {
+    query = query.merge({ provider })
+
+    return this.reql_call(query)
```

Comment:
Note the `query` here returns the relevant row from the `users` table (regardless of which branch above is hit). We then merge in the `provider`, and use that object to sign the JWT.